Easee: validate LIFETIME_ENERGY and SESSION_ENERGY

allcoolusernamesaregone opened the pull request:
As discussed in #9884 , value of LIFETIME_ENERGY and SESSION_ENERGY are often invalid and outdated.

allcoolusernamesaregone committed:
add constant for action poll_lifetimeenergy

allcoolusernamesaregone committed:
Fix wrong and old SESSION_ENERGY values and request updates of LIFETIME_ENERGY

Fix following problems
- After starging a charging session, old values from previous session are delivered by the api. 
- Values of 0 are delivered erratic
- lifetime energy is only updated once per hour and often outdated  when starting and stopping a session

naltatis commented:
@allcoolusernamesaregone please come up with a better PR name that describes whats done here.

andig commented:
/cc @GrimmiMeloni

allcoolusernamesaregone committed:
Merge branch 'master' into master

allcoolusernamesaregone committed:
Add field enabled as added in #9866

Field was lost because of parallel work in the struct

allcoolusernamesaregone commented:
Missing field (added in parallel by @GrimmiMeloni in #9866 )) added in the PR

GrimmiMeloni commented on charger/easee.go:
```diff
@@ -280,6 +281,50 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+
+	switch res.ID {
+		case easee.CHARGER_OP_MODE:
+			//for every op mode change, if last energy updated was triggered more than 3 minutes ago, request new update
```
Do we need this for every opMode change? Can't we apply logic similar toL299 below, e.g. only trigger poll if new opMode is 2 (charging) or 1 (disconnected)?

allcoolusernamesaregone replied:
Yes we could limit to some op modes, but does it hurt to update more frequently, obeying the limit of max 1 call in 3 minutes?
But we could.
With authentication, op mode 7 (awaiting authentication) is one op mode to use as trigger.
No matter if the charging will start immediately or later (depending on the evcc charging mode), we will receive an actual lifetime energy shortly, which will not change until evcc or the user starts the charging.

Also, if EVCC stops the charging (in op mode 3), op mode 7 will be the next op mode to reach.
But if the user stops the charging by unlocking and disconnecting the car, the op mode will change from 3 to 1. 
So **reaching** op mode 1 or 7 should be sufficient, no matter of the previous op mode.
This will update also if evcc pauses the charging (op mode 3->7) - and update the UI, what's ok.

But i'm not sure about boxes without authentication, which op mode will follow on which op mode.
I assume connecting a car will change from 1 to (6 or 2) to 3. Not sure. But no matter what, we could use **leaving** op mode 1, no matter of the next op mode.  Again if the user stops the charging, the op mode should change from 3 to 1, so **leaving** op mode 3 could trigger.

Sumamry, reaching 1 and 7, or leaving 1 and 3, should trigger. What do you think?

GrimmiMeloni commented on charger/easee.go:
```diff
@@ -280,6 +281,50 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+			//for every op mode change, if last energy updated was triggered more than 3 minutes ago, request new update
+	  		if c.lastEnergyPollTriggered.Before(time.Now().Add(-3 * time.Minute)) {
+				uri := fmt.Sprintf("%s/chargers/%s/commands/%s", easee.API, c.charger, easee.PollLifetimeEnergy)
+				c.Post(uri, request.JSONContent, request.MarshalJSON(nil))  //do not wait for success, fire&forget
```
At least error handling should be done, e.g.

```
	if _, err := c.Post(uri, request.JSONContent, request.MarshalJSON(nil)); err != nil {
		//log error
	}
```

allcoolusernamesaregone replied:
ok, will add this..

GrimmiMeloni commented on charger/easee.go:
```diff
@@ -280,6 +281,50 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+				uri := fmt.Sprintf("%s/chargers/%s/commands/%s", easee.API, c.charger, easee.PollLifetimeEnergy)
+				c.Post(uri, request.JSONContent, request.MarshalJSON(nil))  //do not wait for success, fire&forget
+				c.lastEnergyPollTriggered = time.Now();
+				//TODO: wait for async update of LIFETIME_ENERGY?
```
this actually won't be as easy as I thought, as we cannot wait for another ProductUpdate while already processing one. The race condition hunch of mine so far is exactly that - a hunch. Let's try and see if and how often we really hit it. Maybe the problem is purely theoretical.

GrimmiMeloni replied:
I think we can work around this problem by explicitly fetching the Lifecycle Energy via the API.
I just tested this API call here: https://developer.easee.cloud/reference/get_api-sites-siteid-energy

It gives energy data per charger like this:

```
[
  {
    "chargerId": "XXXXXXXX",
    "lifeTimeEnergy": 1875.9724827777784
  }
]
```

Just need to fetch this and match the charger ID to get the correct value. This is a synchronous API, so it will give us exact results. And it solves the race condition problem, too.

allcoolusernamesaregone replied:
> I just tested this API call here: https://developer.easee.cloud/reference/get_api-sites-siteid-energy

I'll have a look into this. But I think/fear this could maybe deliver just an up to an hour old value. As I see it, the poll API polls an actual reading from the box. We have to check "your" API above while charging too see If the value changes with every call.
For the poll -> Productupdate it Changes. Maybe thats the cause for the 3 Minute Limit.

will check and process your other comments tomorrow...

GrimmiMeloni replied:
Thanks for going the extra mile and checking the behavior of the API.

So this would leave us with 
- a poll request after the op_mode changes (to capture data as accurate as possible, especially on disconnect). This is what the current state of the PR provides.
- additionally we would want to poll this regularly, independent of any op_mode changes

Is my understanding correct? If so, will you put in the poll? It would involve using a [ticker](https://gobyexample.com/tickers), and letting it run in a go [routine](https://gobyexample.com/goroutines) launched while we init the charger. This might sound "scary" at first, but given your code here so far, I am pretty sure you can do it. 💪

allcoolusernamesaregone replied:
Yes, that's correct so far.
You mean I can do it.. let's see. BTW, my IDE is vi ;)
I updated the PR, could you please review the ticker + routine stuff, and the updates done to ChargedEnergy and if the new func REquestLifetimeEnergyUpdate is ok.. thanks!

GrimmiMeloni commented on charger/easee.go:
```diff
@@ -280,6 +281,50 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+			}
+
+		case easee.SESSION_ENERGY:
+			/* should not be necessary any longer as
```
Let's remove this then. ;)

GrimmiMeloni commented on charger/easee.go:
```diff
@@ -280,6 +281,50 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+   
+	    	//but this is still needed. SESSION_ENERGY must not be set to 0 by Productupdates, they occur erratic
+		    //Reset to 0 is therefore done some lines above
+	    	if prevTime, ok := c.obsTime[res.ID]; ok && prevTime.Equal(res.Timestamp) && 0 == value.(float64) {
```
Do we need the time check at all then? Based on your comment above, my understanding is that setting to 0 via Product Update is never right - independent of any possible prior update.

allcoolusernamesaregone replied:
I think that's right.. every update to 0 can be ignored.

GrimmiMeloni commented on charger/easee.go:
```diff
@@ -280,6 +281,50 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
 		return
 	}
+
+	switch res.ID {
```
Please just integrate into the larger switch block below.

allcoolusernamesaregone replied:
ok

GrimmiMeloni commented on charger/easee/types.go:
```diff
@@ -10,6 +10,8 @@ const (
 	ChargeResume = "resume_charging"
 )
 
+const PollLifetimeEnergy = "poll_lifetimeenergy"
```
Please pull up into the const block above with the other command strings.

allcoolusernamesaregone committed:
Moved action poll_lifetimeenergy to other action constants

allcoolusernamesaregone committed:
Reworked ProductUpdates of CHARGER_OP_MODE and SESSION_ENERGY

allcoolusernamesaregone commented:
PR reworked as discussed above. Todo: check if https://developer.easee.cloud/reference/get_api-sites-siteid-energy would help us..

allcoolusernamesaregone committed:
Aligned the string values

allcoolusernamesaregone committed:
Trigger poll lifetime only after an actual change of the op mode

If the SignalR connection is renewed all obsservartions will be redelivered no matter if changed to a new value or not. So op mode "changes" while charging from 3 to 3.

allcoolusernamesaregone committed:
Rework CHARGER_OP_MODE, LIFETIME_ENERGY and SESSION_ENERGY productupdates

Add ticker for periodic requests to update lifetime energy.
Use lifetime energy as an alternative way to session energy as value for ChargedEnergy - use the higher (assume: newer) one.

allcoolusernamesaregone committed:
Updated some comments

allcoolusernamesaregone committed:
Fix ticker / stop channel

GrimmiMeloni commented on charger/easee.go:
```diff
@@ -292,9 +296,19 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+		//SESSION_ENERGY must not be set to 0 by Productupdates, they occur erratic
+		//Reset to 0 is done in case CHARGER_OP_MODE
+		if 0 == value.(float64) {
+			c.log.TRACE.Printf("ProductUpdate of SESSION_ENERGY to 0 ignored");
```
We don't need this log, as we log all ProductUpdates above.
As value is never mutated that suffices to understand the flow from the logs (if we ever need to).
As a bonus, you can then also negate the if condition above, and get rid of the return below. Streamlines things.

GrimmiMeloni commented on charger/easee.go:
```diff
@@ -306,6 +320,53 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
 		c.dynamicChargerCurrent = value.(float64)
 	case easee.CHARGER_OP_MODE:
+		//for relevant op mode changes, leaving op mode 1 and 3, or reaching op mode 1 and 7, request new update ofLIFETIME_ENERGY
+		if !(c.opMode == value.(int)) &&  //only if op mode actually changed AND
```
c.opMode != value.(int)

GrimmiMeloni commented on charger/easee.go:
```diff
@@ -306,6 +320,53 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+			c.log.TRACE.Printf("ProductUpdate: Car connected. Reset internal value of SESSION_ENERGY to 0");
+			c.sessionEnergy = 0;
+			//use -1. Using 0 could not distinguish between reset after session and newly started evcc
+			c.sessionStartLifetimeEnergy = -1;
```
nitpick: move -1 into a well named constant. Can get rid of inline comment then.

Edit: That would include using the constant above as well, of course.

GrimmiMeloni commented on charger/easee.go:
```diff
@@ -685,12 +746,36 @@ func (c *Easee) CurrentPower() (float64, error) {
 
 var _ api.ChargeRater = (*Easee)(nil)
 
+func (c *Easee) RequestLifetimeEnergyUpdate() (err error) {
```
As all callers are ignoring the possible error, just remove it from the signature.

GrimmiMeloni commented on charger/easee.go:
```diff
@@ -306,6 +320,53 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+		//OpMode changed fom something to "no car"
+		//if easee.ModeDisconnected == value.(int) && easee.ModeAwaitingStart <= c.opMode {
+		//	c.log.TRACE.Printf("ProductUpdate: Car disconnected")
+		//	maybe needed later
```
lets get rid of "dead" code

GrimmiMeloni reviewed (changes requested):
Only some minor remarks.
Good job with the go routine. 😉 👍 

One minor (or major) concern that keeps at least me from approving is the logging statements. There was previously a deadlock in evcc that was found to be related to logging while holding c.mux. See also inline comment in line 269.

@andig I checked #8009 but could not really determine if the logging that was removed here, or the conflict with influx that you mentioned in another comment was found to be the root cause. So I am unsure if leaving these log statements as they are right now might actually cause a regression of said dead lock. Can you provide guidance?

allcoolusernamesaregone commented:
> Only some minor remarks. Good job with the go routine. 😉 👍
:)

> One minor (or major) concern that keeps at least me from approving is the logging statements. There was previously a deadlock in evcc that was found to be related to logging while holding c.mux. See also inline comment in line 269.
... -> Log output will be removed. It's primary purpose are tests..

allcoolusernamesaregone committed:
Cleanup, tidy up, remove trace output

Tests ok. Lifetime energy updates every 5 minutes are working. Lifetime energy updated by api, calced differences to the remembered start value are showing constant charging progress.
Start and end value are valid when updating session in database, charged amount also.

allcoolusernamesaregone committed:
Add constant NEED_SESSION_START_ENERGY

allcoolusernamesaregone committed:
Merge branch 'evcc-io:master' into master

allcoolusernamesaregone committed:
Merge branch 'master' into master

allcoolusernamesaregone committed:
Minor improvement how SESSION_ENERGY of 0 are ignored

as discussed here
https://github.com/evcc-io/evcc/pull/9940#discussion_r1333550275

allcoolusernamesaregone commented:
So mal auf deutsch da schneller:

Mit dieser Version von ChargedEnergy() (ist im PR jetzt alles hübsch, war nur zu Testzwecken):
```
maxValue := c.sessionEnergy;
if c.sessionStartLifetimeEnergy > 0 {
        calcedEnergy := c.totalEnergy - c.sessionStartLifetimeEnergy; //we could use math.Max, but for now we want the trace output
        if maxValue < calcedEnergy {
                c.log.TRACE.Printf("Replacing SESSION_ENERGY %v by greater calced session energy %v", maxValue, calcedEnergy);
                maxValue = calcedEnergy;
        } else {
                c.log.TRACE.Printf("USING SESSION_ENERGY, its greater %v, %v", maxValue, calcedEnergy)
        }
}
return maxValue, nil
```

ergibt sich dieser Logoutput beim Laden. Man sieht schön, dass nach dem "Bestellen" eines Updates LIFETIME_ENERGY größer wird, und die Differenz zum gemerkten Startwert von LIFETIME_ENERGY aktueller/größer ist als die als Update gelieferten Werte von SESSION_ENERGY - bis dann SESSION_ENERGY aktualisiert wird und das dann wieder größer/neuer ist. 
Das ergibt eine häufige Aktualisierung in der UI über die bisher geladene Energiemenge - ganz abgesehen von dem eigentlichen Ziel eines korrekten Start-/Endwerts sowie geladener Menge in der Datenbank später.

```
[easee ] 11:24:18 POST https://api.easee.com/api/chargers/*/commands/poll_lifetimeenergy
[easee ] 11:24:19 ProductUpdate *: (2023-09-22 09:24:18 +0000 UTC) LIFETIME_ENERGY 2387.2566883333316
[easee ] 11:24:21 Replacing SESSION_ENERGY 1.3620308637619019 by greater calced session energy 1.685906666666142
...
[easee ] 11:30:02 ProductUpdate *: (2023-09-22 09:30:01 +0000 UTC) SESSION_ENERGY 2.5625195503234863
[easee ] 11:30:04 USING SESSION_ENERGY, its greater 2.5625195503234863, 1.685906666666142
...
[easee ] 11:34:18 POST https://api.easee.com/api/chargers/*/commands/poll_lifetimeenergy
[easee ] 11:34:19 ProductUpdate *: (2023-09-22 09:34:18 +0000 UTC) LIFETIME_ENERGY 2388.133301111109
[easee ] 11:34:21 USING SESSION_ENERGY, its greater 2.5625195503234863, 2.562519444443751
...
[easee ] 11:44:18 POST https://api.easee.com/api/chargers/*/commands/poll_lifetimeenergy
[easee ] 11:44:19 ProductUpdate *: (2023-09-22 09:44:18 +0000 UTC) LIFETIME_ENERGY 2388.7718633333316
[easee ] 11:44:22 Replacing SESSION_ENERGY 2.5625195503234863 by greater calced session energy 3.2010816666661412
...
[easee ] 11:50:02 ProductUpdate *: (2023-09-22 09:50:01 +0000 UTC) SESSION_ENERGY 3.6240127086639404
[easee ] 11:50:04 USING SESSION_ENERGY, its greater 3.6240127086639404, 3.2010816666661412

```

andig commented on charger/easee.go:
```diff
@@ -43,28 +44,31 @@ import (
 	cmdC    chan easee.SignalRCommandResponse
 	obsC    chan easee.Observation
 	obsTime map[easee.ObservationID]time.Time
+	ticker		*time.Ticker
```
Bitte gofmt benutzen

andig replied:
Der Ticker entfällt wenn lokal in der Goroutine

andig commented on charger/easee.go:
```diff
@@ -292,9 +297,15 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
 		c.currentPower = 1e3 * value.(float64)
 	case easee.SESSION_ENERGY:
-		c.sessionEnergy = value.(float64)
+		//SESSION_ENERGY must not be set to 0 by Productupdates, they occur erratic. Reset to 0 is done in case CHARGER_OP_MODE
```
Kommentare nach // bitte ein Leerzeichen

andig commented on charger/easee.go:
```diff
@@ -292,9 +297,15 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
 	case easee.SESSION_ENERGY:
-		c.sessionEnergy = value.(float64)
+		//SESSION_ENERGY must not be set to 0 by Productupdates, they occur erratic. Reset to 0 is done in case CHARGER_OP_MODE
+		if 0 != value.(float64) {
```
Yoda question- bitte umdrehen ;)

andig commented on charger/easee.go:
```diff
@@ -292,9 +297,15 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+		}
 	case easee.LIFETIME_ENERGY:
 		c.totalEnergy = value.(float64)
+		if easee.NEED_SESSION_START_ENERGY == c.sessionStartLifetimeEnergy {
```
Yoda

andig commented on charger/easee.go:
```diff
@@ -306,6 +317,46 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+		//New charging session pending, reset internal value of SESSION_ENERGY to 0, and its observation timestamp to "now".
+		//This should be done in a proper way by the api, but it's not.
+		//Remember value of LIFETIME_ENERGY as start value of the charging session
+		if easee.ModeDisconnected >= c.opMode && easee.ModeAwaitingStart <= value.(int) {
```
Yoda

andig commented on charger/easee.go:
```diff
@@ -306,6 +317,46 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+		//Remember value of LIFETIME_ENERGY as start value of the charging session
+		if easee.ModeDisconnected >= c.opMode && easee.ModeAwaitingStart <= value.(int) {
+			c.sessionEnergy = 0
+			loc, _ := time.LoadLocation("UTC")
```
Wofür brauchts die Location?

andig commented on charger/easee.go:
```diff
@@ -306,6 +317,46 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+			c.sessionStartLifetimeEnergy = easee.NEED_SESSION_START_ENERGY
+		}
+		//OpMode changed TO charging. Start ticker for periodic requests to update LIFETIME_ENERGY
+		if easee.ModeCharging == value.(int) && easee.ModeCharging != c.opMode {
```
Yoda

andig commented on charger/easee.go:
```diff
@@ -306,6 +317,46 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+		//OpMode changed TO charging. Start ticker for periodic requests to update LIFETIME_ENERGY
+		if easee.ModeCharging == value.(int) && easee.ModeCharging != c.opMode {
+			c.ticker = time.NewTicker(5 * time.Minute)
+			c.stopTicker = make(chan bool)
```
Goish wäre ein 

    chan struct{}

andig commented on charger/easee.go:
```diff
@@ -306,6 +317,46 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+
+		//OpMode changed FROM charging to something else - stop ticker
+		if easee.ModeCharging == c.opMode && easee.ModeCharging != value.(int) {
+			c.ticker.Stop()
```
Der ticker sollte nur in der Goroutine leben

allcoolusernamesaregone replied:
So, ich hab das probiert, ich kenne nicht die Geheimnisse von Go, daher wider besserer Befürchtung wie folgt getestet:
 
```
	case easee.CHARGER_OP_MODE:
	  
	  ...
	  
		var ticker *time.Ticker
		var stopTicker chan struct{}

		// OpMode changed TO charging. Start ticker for periodic requests to update LIFETIME_ENERGY
		if c.opMode != easee.ModeCharging && value.(int) == easee.ModeCharging {
			ticker = time.NewTicker(5 * time.Minute)
			stopTicker = make(chan struct{})
			go func() {
				for {
					select {
					case <-stopTicker:
						return
					case <-ticker.C:
						c.requestLifetimeEnergyUpdate()
					}
				}
			}()
		}

		// OpMode changed FROM charging to something else - stop ticker
		if c.opMode == easee.ModeCharging && value.(int) != easee.ModeCharging {
// ================
			ticker.Stop()
			stopTicker <- struct{}{}
		}
```

Irgendwie erwartungsgemäß knallt das bei // ================, da ticker beim Aufruf von 
`func (c *Easee) ProductUpdate(i json.RawMessage) { ` 
im `case easee.CHARGER_OP_MODE` und wahrer Bedingung `if c.opMode == easee.ModeCharging && value.(int) != easee.ModeCharging {`   nil ist. 
Wird ja auch nur beim Aufruf der Funktion im Fall `if c.opMode != easee.ModeCharging && value.(int) == easee.ModeCharging {` gefüllt.

`[easee ] TRACE 2023/09/25 11:26:57 level=info connection=lEwEPYssHz2GwWWHgEcZBA event=panic in target method error=runtime error: invalid memory address or nil pointer dereference name=ProductUpdate reaction=send completion with error`

Übersehe ich hier irgendein Konstrukt in Go, das es in C oder Java nicht gibt, um hier den Ticker nur in der Goroutine leben zu lassen..?

Ich habe das also erstmal so gelassen. Der ganze Rest ist aber umgesetzt :)

allcoolusernamesaregone replied:
. oder ist das `ticker.Stop()` völlig unnötig, und `stopTicker <- struct{}{}` ausreichend um den Ticker zu stoppen?
Dann könnte der `ticker `in der Tat wie im Test lokal definiert sein, und nur `stopTicker` global (oder wie auch immer es in Go denn heißt.. ;))

GrimmiMeloni replied:
Ja, so sollte das sein. Wenn Du aus der go routine via return aussteigst, geht der Ticker out-of-scope und wird dann irgendwann von der GC eingesammelt. Du kannst aber auch trotzdem den Ticker noch innerhalb der Go Routine direkt vor dem return anhalten. Vielleicht unnötig - da bin auch ich jetzt überfragt.

andig commented on charger/easee.go:
```diff
@@ -683,13 +734,29 @@ func (c *Easee) CurrentPower() (float64, error) {
 	return c.currentPower, nil
 }
 
+func (c *Easee) RequestLifetimeEnergyUpdate() {
```
```suggestion
func (c *Easee) requestLifetimeEnergyUpdate() {
```

andig commented on charger/easee.go:
```diff
@@ -683,13 +734,29 @@ func (c *Easee) CurrentPower() (float64, error) {
+		if _, err := c.Post(uri, request.JSONContent, request.MarshalJSON(nil)); err != nil {
+			c.log.WARN.Printf("Failed to trigger an update of LIFETIME_ENERGY: %v", err)
+		}
+		c.lastEnergyPollTriggered = time.Now()
```
Brauchts den? Das macht doch schon der Ticker?

GrimmiMeloni replied:
Der Ticker allein triggert alle 5 Minuten, `requestLifetimeEnergyUpdate()` wird aber auch noch ein finales Mal aufgerufen wenn ein Disconnect auftritt, damit die WB Ihre akuellsten Daten rausrückt. Je nach Timing könnten wir durch diesen Aufruf dann aber das 1x pro 3min Threshold auf der API reissen und wir würden 429er erhalten. Deshalb dieser Timestamp.

andig commented:
@allcoolusernamesaregone willst du den PR noch fertig machen?

allcoolusernamesaregone commented:
> willst du den PR noch fertig machen?
Moin, ja klar, hatte nur Wiesn statt Tastatur ;)
Geht spät. morgen hier weiter...!

andig commented:
> Moin, ja klar, hatte nur Wiesn statt Tastatur ;)

Ohja :)

allcoolusernamesaregone committed:
Merge branch 'master' into master

allcoolusernamesaregone committed:
Rework easee.go as discussed in PR

As commented by @andig

GrimmiMeloni reviewed (approved):
Bis auf die Kleinigkeit mit dem Ticker in der routine ist hier jetzt meiner Ansicht nach alles gut.

andig commented on charger/easee.go:
```diff
@@ -306,6 +318,46 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+
+		// OpMode changed TO charging. Start ticker for periodic requests to update LIFETIME_ENERGY
+		if c.opMode != easee.ModeCharging && value.(int) == easee.ModeCharging {
+			c.ticker = time.NewTicker(5 * time.Minute)
```
Der Ticket gehört m.E. _in_ die Goroutine...

allcoolusernamesaregone replied:
done

andig commented on charger/easee.go:
```diff
@@ -683,13 +735,29 @@ func (c *Easee) CurrentPower() (float64, error) {
+	// or the SESSION_ENERGY value by the API. Each value could be lower than the other, depending on
+	// order and receive timestamp of the product update. We want to return the higher (and newer) value.
+	if c.sessionStartLifetimeEnergy > 0 {
+		return math.Max(c.sessionEnergy, c.totalEnergy-c.sessionStartLifetimeEnergy), nil
```
Da sollte `max` reichen

allcoolusernamesaregone replied:
> Da sollte `max` reichen

done

allcoolusernamesaregone committed:
Rework minor comments

Moved ticker to goroutine
Removed math package

allcoolusernamesaregone commented:
Das Updateintervall von 5 Minuten lässt sich ggf. auch noch optimieren, bzw. Aufrufe reduzieren.
Aktuell kommt nur nach jedem 2. Aufruf ein neuer Wert, also alle 10 Minuten. 
Ggf. ist ein höheres Intervall von 6 oder 8 Minuten sogar besser, wenn dafür jedes Mal, und somit im Ergebnis öfter, ein aktualisierter Wert zurückkäme bzw. das Productupdate dazu. 
Ich muss aber noch prüfen, ob das ggf. von der Ladestärke abhängt bzw. irgendeine Korrelation zwischen Ladestärke und Updatehäufigkeit vorliegt.

So oder so, der PR ist mehrfach in verschiedenen Szenarien von Start und Stopp erfolgreich getestet. Es darf aber gerne sonst noch jemand ;)

![Updates_LIFETIME_ENERGY](https://github.com/evcc-io/evcc/assets/22935408/1fdea6f8-1f08-4cbb-95d7-b28015f8a7de)

allcoolusernamesaregone committed:
Merge branch 'evcc-io:master' into master

allcoolusernamesaregone committed:
Merge branch 'evcc-io:master' into master

allcoolusernamesaregone committed:
Merge branch 'evcc-io:master' into master

andig commented on charger/easee.go:
```diff
@@ -306,6 +316,45 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+
+		// OpMode changed TO charging. Start ticker for periodic requests to update LIFETIME_ENERGY
+		if c.opMode != easee.ModeCharging && value.(int) == easee.ModeCharging {
+			c.stopTicker = make(chan struct{})
```
Ist sicher gestellt, dass der Channel hier immer nil ist bevor wir das machen? Mal probehalber mit einem panic versehen? Anderenfalls wäre das ein leak.

GrimmiMeloni replied:
@andig Ich hätte jetzt erwartet das mit Überschreiben von c.stopTicker ein möglicher vorhandener Channel einfach out of scope geht, und dann von der GC erfasst wird. Ist das ggf. bei Channels anders?

andig replied:
Wenn er nicht closed ist läuft die Goroutine weiter und die verwendet den Channel. Wie stellen wir sicher, dass der Channel immer closed (oder geschrieben) ist bevor wir einen neuen schreiben?

GrimmiMeloni replied:
Danke! Jetzt sehe ich zumindest das Problem. Wenn wir OP_MODE updates out of order bekommen könnten wir hier landen bevor in 346 die Routine anhalten. Good catch!

Zur Lösung - könnten wir hier nicht einfach einen nil check einbauen, und nur bei != nil den Channel erstellen und die GoRoutine starten? Dazu dann noch wie unten angefordert in L346++ den Channel schließen und den Member auf nil setzen?

andig replied:
Genau. Dann unten close und auf nil setzen, aber mit Mutex.

allcoolusernamesaregone replied:
Sollte nach meinem Verständnis nun hübsch sein.
Zusätzlich habe ich in Zeile 351 ein `c.stopTicker != nil` angehängt, denn denke mit der gleichen Logik, warum der channel hier noch != nil sein könnte
```
// OpMode changed TO charging. Start ticker for periodic requests to update LIFETIME_ENERGY
		if c.opMode != easee.ModeCharging && value.(int) == easee.ModeCharging {
			if c.stopTicker == nil {
				c.stopTicker = make(chan struct{})
			}
```
könnte er hier ja auch **noch nicht** != nil sein.
```
		// OpMode changed FROM charging to something else - stop ticker if channel exists
		if c.opMode == easee.ModeCharging && value.(int) != easee.ModeCharging && c.stopTicker != nil {
			c.stopTicker <- struct{}{}
			close(c.stopTicker)
			c.stopTicker = nil
		}

```

Wobei da die ProductUpdates aber wirklich sehr extrem durcheinander kommen müssten, insb. in Anbetracht des Filters auf das Alter. Aber schaden tut's ja nicht...

andig commented on charger/easee.go:
```diff
@@ -306,6 +316,45 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+
+		// OpMode changed FROM charging to something else - stop ticker
+		if c.opMode == easee.ModeCharging && value.(int) != easee.ModeCharging {
+			c.stopTicker <- struct{}{}
```
This should close the channel.

andig commented on charger/easee.go:
```diff
@@ -680,13 +729,29 @@ func (c *Easee) CurrentPower() (float64, error) {
 }
 
+func (c *Easee) requestLifetimeEnergyUpdate() {
+	if c.lastEnergyPollTriggered.Before(time.Now().Add(-3 * time.Minute)) { //api rate limit, max once in 3 minutes
```
if time.Since …

allcoolusernamesaregone replied:
geändert

andig replied:
Musst du nicht kommentieren- wenn Code sich ändern zeigt Github das direkt zu den Kommentaren an.

andig commented:
@GrimmiMeloni willst Du Dir das anschauen? Ich würde auch gerne `NEED_SESSION_START_ENERGY` noch zugunsten eines `*float64` los werden.

allcoolusernamesaregone commented:
> @GrimmiMeloni willst Du Dir das anschauen? Ich würde auch gerne `NEED_SESSION_START_ENERGY` noch zugunsten eines `*float64` los werden.

ja das wäre gut wenn hier den Feinschliff dann jemand mit Background machen könnte, habe mit Go und EVCC nicht wirklich bzw. erst seit dem PR hier zu tun...  :)

allcoolusernamesaregone committed:
Merge branch 'master' into master

allcoolusernamesaregone committed:
Merge branch 'master' into master

allcoolusernamesaregone committed:
Update easee.go

Create channel only if it does not exist
Use channel (to stop ticker) only if it exists
Close channel and nil it after sending stop signal

allcoolusernamesaregone committed:
Update easee.go

🙄

allcoolusernamesaregone commented:
> @GrimmiMeloni willst Du Dir das anschauen? Ich würde auch gerne `NEED_SESSION_START_ENERGY` noch zugunsten eines `*float64` los werden.

Verstehe ich nicht ganz, was genau meinst Du?

andig committed:
Fix creating multiple go routines

andig commented on charger/easee.go:
```diff
@@ -305,8 +315,58 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+
+		// OpMode changed TO charging. Start ticker for periodic requests to update LIFETIME_ENERGY
+		if c.opMode != easee.ModeCharging && opMode == easee.ModeCharging {
+			if c.stopTicker == nil {
```
So geändert, dass wir keine neuen go routinen erzeigen wenn die alte noch da ist.

andig commented on charger/easee.go:
```diff
@@ -305,8 +315,58 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+
+		// OpMode changed FROM charging to something else - stop ticker if channel exists
+		if c.opMode == easee.ModeCharging && opMode != easee.ModeCharging && c.stopTicker != nil {
+			close(c.stopTicker)
```
`close` triggert den Channel- es ist nicht notwendig, zusätzlich noch etwas zu senden.

andig commented:
Ich hab mir das jetzt lange angeschaut und glaube, dass der Code falsch ist: ProductUpdate locket die Mutex, bei opmode Änderung machen wir das plötzlich nochmal. Das müsste eigentlich zu einem Deadlock führen. Warum kann das überhautp klappen?

Zweiter Fehler: der Zugriff auf energyPollTime ist racy und muß ebenfalls per Mutex geschützt werden.

GrimmiMeloni committed:
address comments around locking

andig commented:
@GrimmiMeloni was mich viel mehr stört: der Code oben wurde ja getestet. Nur wie? Das *kann* nicht funktioniert haben! Wäre auch schön, das ohne weitere Mutex zu lösen- eine reicht.

GrimmiMeloni commented:
> Ich hab mir das jetzt lange angeschaut und glaube, dass der Code falsch ist: ProductUpdate locket die Mutex, bei opmode Änderung machen wir das plötzlich nochmal. Das müsste eigentlich zu einem Deadlock führen. Warum kann das überhautp klappen?

Gar nicht. Aber den Deadlock hat @allcoolusernamesaregone auch erst gestern eingebaut. Vermutlich einfach noch nicht getestet.

> Zweiter Fehler: der Zugriff auf energyPollTime ist racy und muß ebenfalls per Mutex geschützt werden.

Habe ich gerade eingebaut.

andig commented on charger/easee.go:
```diff
@@ -305,8 +316,55 @@ func (c *Easee) ProductUpdate(i json.RawMessage) {
+		if c.opMode != opMode && // only if op mode actually changed AND
+			(c.opMode == easee.ModeDisconnected || c.opMode == easee.ModeCharging || // from these op modes
+				opMode == easee.ModeDisconnected || opMode == easee.ModeAwaitingAuthentication) { // or to these op modes
+			c.requestLifetimeEnergyUpdate()
```
Das würde jetzt auch für den Call blocken wenn eine OpMode Änderung kommt während das Energy Update läuft. Wollen wir das?

GrimmiMeloni replied:
Ja, das ist OK. Wir warten bei dem request für den Poll nicht auf eine Antwort. Die Wartezeit beschränkt sich hier auf den reinen HTTP Roundtrip. Der ist (selbst bei Easee 🤡 ) immer nach ein paar Millis erledigt.

andig replied:
...jedenfalls solange wir dem Ding auch eine eigene Mutex spendieren. Sonst blockiert das den SignalR. Das wäre dann wieder ein Argument dafür.

andig commented:
> Gar nicht. Aber den Deadlock hat @allcoolusernamesaregone auch erst gestern eingebaut. Vermutlich einfach noch nicht getestet.

Ok, das beruhigt :)

GrimmiMeloni commented:
> Wäre auch schön, das ohne weitere Mutex zu lösen- eine reicht.

Der Aufruf kommt entweder aus dem ProductUpdate Handler (der hält mux.Lock())
oder aus der Go Routine (die hält max.Lock() NICHT).

Mir schien der dedizierte Mux einfacher/straight forward. Wir können aber auch über einen Boolean (oder eine zweite Wrapper Funktion) das ganze steuern. @andig hast Du bei solchen Situationen Präferenzen?

GrimmiMeloni commented:
> @GrimmiMeloni willst Du Dir das anschauen? Ich würde auch gerne `NEED_SESSION_START_ENERGY` noch zugunsten eines `*float64` los werden.

Geht es Dir um die saubere Typisierung des Const (int vs float) und Du willst die implizite Typwandlung loswerden?

GrimmiMeloni committed:
fix linter, correctly type const

andig commented:
Die Konstante zeigt ja nur „kein Wert“. Das können wir auch durch einen float pointer = nil machen. Dann brauchts die Konstante nicht mehr.

allcoolusernamesaregone commented:
> > Gar nicht. Aber den Deadlock hat @allcoolusernamesaregone auch erst gestern eingebaut. Vermutlich einfach noch nicht getestet.
> 
> Ok, das beruhigt :)

Ja, sorry, ich habe einfach treudoof wie gewünscht den Mutex da eingebaut, und auch nicht drüber nachgedacht das nochmal zu testen. Aber, aktuell mit der von Euch geänderten Version (mit Commit [0d898f4](https://github.com/evcc-io/evcc/commit/0d898f40edde0ee57f68dff4d9ea63751710469e)) - Ladevorgang läuft sauber..

allcoolusernamesaregone committed:
Merge branch 'evcc-io:master' into master

andig committed:
Remove const

andig commented:
I have removed the const- could you check once more?

allcoolusernamesaregone committed:
Merge branch 'evcc-io:master' into master

allcoolusernamesaregone committed:
Merge branch 'evcc-io:master' into master

allcoolusernamesaregone committed:
Merge branch 'evcc-io:master' into master

allcoolusernamesaregone committed:
Merge branch 'evcc-io:master' into master

allcoolusernamesaregone committed:
Merge branch 'evcc-io:master' into master

andig commented:
@GrimmiMeloni good to go?

GrimmiMeloni commented:
> @GrimmiMeloni good to go?

Yes. LGTM. Charged multiple times with this change. No issues.
